### PR TITLE
feat: convert to use recursive types for path.

### PR DESCRIPTION
**Pull request by jspears**
Converted Path to use recursive types, that also use Path notation.

This PR -
1 - removes all Path{1,2,3,4,5} and goes to using recursive types.
2 - makes all paths reachable via `from.here.to.there` notation.

Things to consider
1 -  Make path recursive so '.' notation works inside of arrays.
2 - Create a more sophisticated grammer to allow for array selection, "elems()" think `from.*.to.there`
3 - Allow Pick and other conversion functions to use dot notation to create new Lens's.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -146,8 +146,11 @@ const bar = O.optic<Data>().prop('foo').prop('bar')
 const bar = foo.prop('bar')
 // or use .path() to compose multiple prop lenses with a single call
 const bar = O.optic<Data>().path(['foo', 'bar'])
+// or use path with a nested string
+const bar = O.optic<Data>().path('foo.bar')
 ```
```
Comment (jspears) [inline README.md:151]: changed from overloading prop to overloading path.   Although really prop could go away.

File: package.json [hunk 2]
```diff
@@ -34,8 +34,8 @@
     "doctoc": "^1.4.0",
     "jest": "^26.0.1",
     "prettier": "^2.0.2",
-    "ts-jest": "^26.0.0",
+    "ts-jest": "^26.4.1",
     "ts-node": "^8.8.1",
-    "typescript": "^3.8.3"
+    "typescript": "beta"
   }
```
Comment (jspears) [inline package.json:40]: Just a reminder this requires Typescript 4.1 which is still beta

**Comment by jspears** [comment]
So I think the beta of typescript is a little touchy about formatting.

File: package.json (hunk 2 above)
Comment (jspears) [reply package.json:40]: Perhaps publish a beta version of this library?

File: .prettierignore [hunk 3]
```diff
@@ -0,0 +1 @@
+**/src/descend.ts
```
Comment (jspears) [inline .prettierignore:1]: prettier does not support string literals yet, so I put them all in this typesript file and ignored it.

File: .prettierrc.json [hunk 4]
```diff
@@ -2,6 +2,6 @@
   "arrowParens": "avoid",
   "trailingComma": "es5",
   "tabWidth": 2,
-  "semi": false,
+  "semi": true,
   "singleQuote": true
```
Comment (jspears) [inline .prettierrc.json:6]: Wierd behaviour with typescript 4.1-beta -- don't think the beta version was parsing things correctly.  Turning on semi's made it more consistent.

File: package.json [hunk 5]
```diff
@@ -21,10 +21,11 @@
     "dist"
   ],
   "scripts": {
-    "build": "tsc -p tsconfig.cjs.json && tsc -p tsconfig.esm.json",
+    "build": "./scripts/generate-index.ts > src/index.ts && tsc -p tsconfig.cjs.json && tsc -p tsconfig.esm.json && prettier --write {src,tests}",
+
```
Comment (jspears) [inline package.json:25]: I think making build do all the things is safer.   I didn't realize at first that the real code was in the generate-index, I know the first line of the file says....

File: package.json [hunk 6]
```diff
@@ -21,10 +21,11 @@
...
     "test": "jest",
     "clean": "rm -rf dist/",
-    "prepublishOnly": "yarn run clean && yarn run build",
+    "prepare": "yarn run clean && yarn run build",
     "doctoc": "doctoc README.md --title '## Table of Contents' && prettier --write README.md",
```
Comment (jspears) [inline package.json:29]: [prepublishOnly](https://github.com/npm/cli/pull/268/files) is deprecated changed to prepare

File: src/descend.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { IsEmpty } from './utils'
```
Comment (jspears) [inline src/descend.ts:1]: Added this file so I could exclude it from prettier -- as prettier does not like the typescript syntax.

File: .prettierrc.json (hunk 4 above)
Comment (jspears) [reply .prettierrc.json:6]: semi makes a bit of a mess... if it is a problem I could move the build system changes to a different PR, then do one with semi colons, then do this one.   I not though.

File: package.json (hunk 6 above)
Comment (akheron) [reply package.json:29]: Isn’t prepare also run on initial npm install? It’s useless to build at that point.

File: .prettierrc.json (hunk 4 above)
Comment (akheron) [reply .prettierrc.json:6]: Or maybe wait until the problem is fixed in prettier?

Comment (jspears) [reply .prettierrc.json:6]: Problem isn't really with prettier -- when the semi is off typescript fails to build.

File: package.json (hunk 6 above)
Comment (jspears) [reply package.json:29]: It is run on local install (inside the package) it is not run when somebody installs the package
i.e. `npm add`

Comment (akheron) [reply package.json:29]: Yes, that’s right. I still don’t like it behaving that way. If prepublishOnly is deprecated, do you know if there will be a replacement that works in a similar way?

File: .prettierrc.json (hunk 4 above)
Comment (akheron) [reply .prettierrc.json:6]: Hmm, interesting. So it’s either a bug in TypeScript, or prettier cannot add a semi in some place where it’s actually required with the new TS.

File: package.json (hunk 6 above)
Comment (jspears) [reply package.json]: Nothing exactly the same.  Prepare is the closest.

**Comment by akheron** [comment]
Thanks! 

I still think it’s best to wait for TS 4.1 to be released and the semicolon issue to be resolved before merging this.

Anyway, good work! I hope you’re okay if we just hold our horses a bit longer :)

**Comment by akheron** [comment]
I rewrote this in #15

**Pull request closed without merging**